### PR TITLE
DOC-5184 - cobalt cache enhancement properties

### PR DESCRIPTION
**Pull request by jamesnocentini**
https://issues.couchbase.com/browse/DOC-5184

This PR originates from the [Technical Spec](https://docs.google.com/document/d/1v60oiWY-M0qExEnWb-bvEodj76ds_qfJXoL3U_9locA/edit?ts=5cc1af33#heading=h.ry6bjo56d7j0)

@adamcfraser @bbrks @JRascagneres This PR is updating the schema. Feel free to comment if you are you aware of any other config changes for Cobalt.

### CODE REVIEW

File: modules/ROOT/assets/attachments/sg.yaml [hunk 1]
```diff
@@ -167,32 +167,50 @@ properties:
...
+                max_wait_skipped:
+                enable_star_channel_cache:
+                max_length:
+                min_length:
+                expiry:
+            rev_cache:
+              type: object
+              description: Revision cache configuration
+              properties:
+                size:
```
Comment (bbrks) [inline modules/ROOT/assets/attachments/sg.yaml:244]:
```suggestion
                size:
                  type: integer
                  description: Size of the revision cache, specified as the total number of document revisions to cache in memory for all recently accessed documents. When the revision cache is full, Sync Gateway removes less recent document revisions to make room for new document revisions. Adjust this property to tune memory consumption by Sync Gateway, for example on servers with less memory and in cases when Sync Gateway creates many new documents and/or updates many documents relative to the number of read operations.
                  default: 5000
```

File: modules/ROOT/assets/attachments/sg.yaml [hunk 2]
```diff
@@ -167,32 +167,50 @@ properties:
...
+                shard_count:
```
Comment (bbrks) [inline modules/ROOT/assets/attachments/sg.yaml:248]:
```suggestion
                shard_count:
                  type: integer
                  description: Number of shards the rev cache should be split into. More shards allows for lower cache contention when accessing distinct revisions, at the cost of some memory overhead per-shard. This generally should not greatly exceed the number of CPU threads available to Sync Gateway.
                  default: 8
```

File: modules/ROOT/assets/attachments/sg.yaml [hunk 3]
```diff
@@ -167,32 +167,50 @@ properties:
...
               channel_cache_expiry:
                 type: integer
-                description: Time (seconds) to keep entries in cache beyond the minimum retained.
+                description: (Deprecated) Time (seconds) to keep entries in cache beyond the minimum retained.
                 default: 60
+            channel_cache:
+              type: object
+              description: Channel cache configuration
+              properties:
+                max_number:
```
Comment (bbrks) [inline modules/ROOT/assets/attachments/sg.yaml:208]:
```suggestion
                max_number:
                  type: integer
                  description: Maximum number of channel caches which will exist at any one point
                  default: 50000
```

File: modules/ROOT/assets/attachments/sg.yaml [hunk 4]
```diff
@@ -167,32 +167,50 @@ properties:
...
+                max_wait_pending:
```
Comment (bbrks) [inline modules/ROOT/assets/attachments/sg.yaml:212]:
```suggestion
                max_wait_pending:
                  type: integer
                  description: Maximum wait time in milliseconds for a pending sequence before skipping sequences.
                  default: 5000
```

File: modules/ROOT/assets/attachments/sg.yaml [hunk 5]
```diff
@@ -167,32 +167,50 @@ properties:
...
+                max_num_pending:
```
Comment (bbrks) [inline modules/ROOT/assets/attachments/sg.yaml:216]:
```suggestion
                max_num_pending:
                  type: integer
                  description: Maximum number of pending sequences before skipping the sequence.
                  default: 10000
```

File: modules/ROOT/assets/attachments/sg.yaml [hunk 6]
```diff
@@ -167,32 +167,50 @@ properties:
...
+                max_wait_skipped:
```
Comment (bbrks) [inline modules/ROOT/assets/attachments/sg.yaml:220]:
```suggestion
                max_wait_skipped:
                  type: integer
                  description: Maximum wait time in milliseconds for a skipped sequence before abandoning the sequence.
                  default: 3600000
```

File: modules/ROOT/assets/attachments/sg.yaml [hunk 7]
```diff
@@ -167,32 +167,50 @@ properties:
...
+                enable_star_channel_cache:
```
Comment (bbrks) [inline modules/ROOT/assets/attachments/sg.yaml]:
```suggestion
                enable_star_channel:
                  type: boolean
                  description: Enable the star (*) channel.
                  default: 'true'
```

File: modules/ROOT/assets/attachments/sg.yaml [hunk 8]
```diff
@@ -167,32 +167,50 @@ properties:
...
+                max_length:
+                min_length:
+                expiry:
```
Comment (bbrks) [inline modules/ROOT/assets/attachments/sg.yaml]:
```suggestion
                expiry_seconds:
                  type: integer
                  description: Time (seconds) to keep entries in cache beyond the minimum retained.
                  default: 60
```

File: modules/ROOT/assets/attachments/sg.yaml [hunk 9]
```diff
@@ -167,32 +167,50 @@ properties:
...
+              type: object
+              description: Channel cache configuration
+              properties:
+                max_number:
+                max_wait_pending:
+                max_num_pending:
+                max_wait_skipped:
+                enable_star_channel_cache:
+                max_length:
+                min_length:
```
Comment (bbrks) [inline modules/ROOT/assets/attachments/sg.yaml:232]:
```suggestion
                min_length:
                  type: integer
                  description: Minimum number of entries maintained in cache per channel.
                  default: 50
```

File: modules/ROOT/assets/attachments/sg.yaml [hunk 10]
```diff
@@ -167,32 +167,50 @@ properties:
...
+            channel_cache:
+              type: object
+              description: Channel cache configuration
+              properties:
+                max_number:
+                max_wait_pending:
+                max_num_pending:
+                max_wait_skipped:
+                enable_star_channel_cache:
+                max_length:
```
Comment (bbrks) [inline modules/ROOT/assets/attachments/sg.yaml:228]:
```suggestion
                max_length:
                  type: integer
                  description: Maximum number of entries maintained in cache per channel.
                  default: 500
```

File: modules/ROOT/assets/attachments/sg.yaml [hunk 11]
```diff
@@ -167,32 +167,50 @@ properties:
...
                 default: 50
               channel_cache_expiry:
                 type: integer
-                description: Time (seconds) to keep entries in cache beyond the minimum retained.
+                description: (Deprecated) Time (seconds) to keep entries in cache beyond the minimum retained.
                 default: 60
+            channel_cache:
+              type: object
+              description: Channel cache configuration
+              properties:
```
Comment (bbrks) [inline modules/ROOT/assets/attachments/sg.yaml:199]: Added high/lower watermark options not in the config tech spec:

```suggestion
              properties:
                compact_high_watermark_pct:
                  type: integer
                  description: High watermark for channel cache eviction (percent)
                  default: 80
                compact_low_watermark_pct:
                  type: integer
                  description: Low watermark for channel cache eviction (percent)
                  default: 60
```

**Review by bbrks — APPROVED** [review]
I've copied the descriptions of the old config options over to the new ones, and made some small modifications based on changes that were made during development.

**Review by bbrks — CHANGES_REQUESTED** [review]
Oops. Hit wrong button for my review above.

**Comment by jamesnocentini** [comment]
@bbrks Perfect!

I'll merge the PR to get it on https://docs-staging.couchbase.com. And we'll set up other rounds of reviews before the release.

**Pull request merged**